common/types/encoding/codecv1/codecv1.go

```diff
@@ -0,0 +1,497 @@
+				hashBytes, err := hex.DecodeString(txHash)
+				if err != nil {
+					return common.Hash{}, err
+				}
+				if len(hashBytes) != 32 {
+					return common.Hash{}, fmt.Errorf("unexpected hash: %s", txData.TxHash)
+				}
+				dataBytes = append(dataBytes, hashBytes...)
+			}
+		}
+	}
+
+	hash := crypto.Keccak256Hash(dataBytes)
+	return hash, nil
+}
+
+// NewDABatch creates a DABatch from the provided encoding.Batch.
+func NewDABatch(batch *encoding.Batch) (*DABatch, error) {
+	// this encoding can only support up to 15 chunks per batch
+	if len(batch.Chunks) > 15 {
+		return nil, fmt.Errorf("too many chunks in batch")
+	}
+
+	// batch data hash
+	dataHash, err := computeBatchDataHash(batch.Chunks, batch.TotalL1MessagePoppedBefore)
+	if err != nil {
+		return nil, err
+	}
+
+	// skipped L1 messages bitmap
+	bitmapBytes, totalL1MessagePoppedAfter, err := constructSkippedBitmap(batch.Index, batch.Chunks, batch.TotalL1MessagePoppedBefore)
+	if err != nil {
+		return nil, err
+	}
+
+	// blob payload
+	blob, z, err := constructBlobPayload(batch.Chunks)
+	if err != nil {
+		return nil, err
+	}
+
+	// blob versioned hash
+	c, err := kzg4844.BlobToCommitment(*blob)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create blob commitment")
+	}
+	blobVersionedHash := kzg4844.CalcBlobHashV1(sha256.New(), &c)
+
+	daBatch := DABatch{
+		Version:                CodecV1Version,
+		BatchIndex:             batch.Index,
+		L1MessagePopped:        totalL1MessagePoppedAfter - batch.TotalL1MessagePoppedBefore,
+		TotalL1MessagePopped:   totalL1MessagePoppedAfter,
+		DataHash:               dataHash,
+		BlobVersionedHash:      blobVersionedHash,
+		ParentBatchHash:        batch.ParentBatchHash,
+		SkippedL1MessageBitmap: bitmapBytes,
+		blob:                   blob,
+		z:                      z,
+	}
+
+	return &daBatch, nil
+}
+
+// computeBatchDataHash computes the data hash of the batch.
+// Note: The batch hash and batch data hash are two different hashes,
+// the former is used for identifying a badge in the contracts,
+// the latter is used in the public input to the provers.
+func computeBatchDataHash(chunks []*encoding.Chunk, totalL1MessagePoppedBefore uint64) (common.Hash, error) {
+	var dataBytes []byte
+	totalL1MessagePoppedBeforeChunk := totalL1MessagePoppedBefore
+
+	for _, chunk := range chunks {
+		daChunk, err := NewDAChunk(chunk, totalL1MessagePoppedBeforeChunk)
+		if err != nil {
+			return common.Hash{}, err
+		}
+		totalL1MessagePoppedBeforeChunk += chunk.NumL1Messages(totalL1MessagePoppedBeforeChunk)
+		chunkHash, err := daChunk.Hash()
+		if err != nil {
+			return common.Hash{}, err
+		}
+		dataBytes = append(dataBytes, chunkHash.Bytes()...)
+	}
+
+	dataHash := crypto.Keccak256Hash(dataBytes)
+	return dataHash, nil
+}
+
+// constructSkippedBitmap constructs skipped L1 message bitmap of the batch.
+func constructSkippedBitmap(batchIndex uint64, chunks []*encoding.Chunk, totalL1MessagePoppedBefore uint64) ([]byte, uint64, error) {
```

Comment:
this should be a common function used by both codec v0 and v1 right?

Reply:
Fixed in [9133ad4](https://github.com/scroll-tech/scroll/pull/1193/commits/9133ad422d9aa11c4ad35c8603e132bfe08488e8).